### PR TITLE
Fixed extraneous non-existing variation images in product teasers.

### PR DESCRIPTION
**Pull request by Mauricio-Urrego**
### Ticket
- [Show arrows only when product variation have seperate images](https://app.asana.com/0/0/1200029023488130/f)

### Description
- Before and after screenshots below.
Screen Shot 2021-04-21 at 21 05 17
Screen Shot 2021-04-21 at 21 04 34

### CODE REVIEW

File: src/WooCommerce.php [hunk 1]
```diff
@@ -342,17 +342,19 @@ public static function woocommerce_template_loop_product_thumbnail() {
...
-            $attachment_ids,
-            $wpdb->get_col(
-              $wpdb->prepare(
-                "SELECT pm.meta_value AS attachment_id
-                FROM wp_posts p
-                INNER JOIN wp_postmeta pm ON pm.post_id = p.ID AND pm.meta_key = '_thumbnail_id'
-                WHERE p.ID IN ($placeholders)
-                ORDER BY p.menu_order ASC",
-                $variation_ids
+        $attachment_ids = array_filter(
```
Comment (wrux) [inline src/WooCommerce.php:345]: This is pulling in data and also doing things that could probably be performed in the query. Instead of using `array_unique()` in PHP, can the query be optimised somehow?

I'm not sure on how the data is structured in the DB but can it use SELECT DISTINCT? 

Would `SELECT DISTINCT pm.meta_value AS attachment_id` work?

Comment (Mauricio-Urrego) [reply src/WooCommerce.php:345]: Yes this works.

**Comment by Mauricio-Urrego** [comment]
Array filter and array unique moved inside sql query as where > 0 and select distinct respectively.

File: src/WooCommerce.php [hunk 2]
```diff
@@ -342,18 +342,16 @@ public static function woocommerce_template_loop_product_thumbnail() {
 
       if (count($variation_ids)) {
         $placeholders = implode(',', array_fill(0, count($variation_ids), '%d'));
-        $attachment_ids = array_unique(
```
Comment (sun) [inline src/WooCommerce.php:345]: The new query looks good, but we still need the array_unique() here to avoid the product's primary image to be repeated, in case a variant uses the same image.

Alternatively, we can omit the array_unique() and instead filter out the primary image ID we already have in the query — see below.

File: src/WooCommerce.php [hunk 3]
```diff
@@ -342,18 +342,16 @@ public static function woocommerce_template_loop_product_thumbnail() {
...
+        $attachment_ids = array_merge(
+          $attachment_ids,
+          $wpdb->get_col(
+            $wpdb->prepare(
+              "SELECT DISTINCT pm.meta_value AS attachment_id
+              FROM wp_posts p
+              INNER JOIN wp_postmeta pm ON pm.post_id = p.ID AND pm.meta_key = '_thumbnail_id'
+              WHERE p.ID IN ($placeholders) AND pm.meta_value > 0
+              ORDER BY p.menu_order ASC",
+              $variation_ids
```
Comment (sun) [inline src/WooCommerce.php:354]:
```suggestion
              WHERE p.ID IN ($placeholders) AND pm.meta_value > 0 AND pm.meta_value <> %d
              ORDER BY p.menu_order ASC",
              $variation_ids, $attachment_ids[0]
```

**Review by sun — CHANGES_REQUESTED** [review]

File: src/WooCommerce.php (hunk 3 above)
Comment (Mauricio-Urrego) [reply src/WooCommerce.php:354]: This is more difficult to read. I'll add back the array_unique if that's okay.

Comment (Mauricio-Urrego) [reply src/WooCommerce.php:354]: @wrux suggested removing array unique and replacing with select distinct https://github.com/netzstrategen/wordpress-gallerya/pull/74#discussion_r618217776

Comment (Mauricio-Urrego) [reply src/WooCommerce.php:354]: Ah I see now ok.

Comment (Mauricio-Urrego) [reply src/WooCommerce.php:354]: Ok well in the interest of preferring everything to be done in the query I suppose this is fine. Will test this.

Comment (Mauricio-Urrego) [reply src/WooCommerce.php:354]: @sun This did not test well. Adding back array_unique. Should I reverse the select distinct change then as well?

**Review by sun — APPROVED** [review]

**Comment by Mauricio-Urrego** [comment]
@sun I'm not sure how I feel about the title change. A bit wordy.

**Comment by fabianmarz** [comment]
I agree.  I have no idea what is meant by it 😂
